### PR TITLE
upgrades cucumber_rails and friends and adjust env to properly exit on fail

### PR DESCRIPTION
**Pull request by tansaku**
Addresses: https://www.pivotaltracker.com/story/show/119350175

### CODE REVIEW

**Comment by tansaku** [comment]
at least this shows the build actually failing - which is what we want ...

File: features/step_definitions/volunteer_op_steps.rb [hunk 1]
```diff
@@ -37,7 +37,7 @@
 end
 
 Then(/^there should be (\d+) doit volunteer ops stored$/) do |count|
-  expect(VolunteerOp.count(source: 'doit')).to eq count.to_i
+  expect(VolunteerOp.count_by_sql("SELECT COUNT(*) FROM volunteer_ops WHERE source = 'doit'")).to eq count.to_i
```
Comment (tansaku) [inline features/step_definitions/volunteer_op_steps.rb]: I'm kind of surprised about this - seems like it's associated with upgrade to Rails 4.2.7? weird though that a version of count would stop working from one minor version to the next - maybe our old cucumber version was using a much older version of rails AR?

File: features/support/env.rb [hunk 2]
```diff
@@ -154,3 +154,7 @@
   # Tell Cucumber to quit after this scenario is done - if it failed.
   Cucumber.wants_to_quit = true if s.failed?
 end
+
+at_exit do
+  exit 1 if Cucumber.wants_to_quit
+end
```
Comment (johnnymo87) [inline features/support/env.rb]: 👍

**Comment by johnnymo87** [comment]
+1 for doing the hard work of fixing a silently failing CI. 

I'm out of the loop, so if you don't mind the questions ...

The change to `features/support/env.rb` is to support a fail-fast option we have configured for cucumber?

The changes to `Gemfile.lock` and `features/step_definitions/volunteer_op_steps.rb` are required to make CI pass?

**Comment by tansaku** [comment]
so weirdly we're now having jasmine fail on this branch:

```
The command "bundle exec rake test_with_coveralls" exited with 0.
3.82s$ USE_JASMINE_RAKE=true bundle exec rake jasmine:ci
[2016-08-10 01:01:42] INFO  WEBrick 1.3.1
[2016-08-10 01:01:42] INFO  ruby 2.3.0 (2015-12-25) [x86_64-linux]
[2016-08-10 01:01:42] INFO  WEBrick::HTTPServer#start: pid=20222 port=45592
rake aborted!
Errno::EINVAL: Invalid argument - connect(2) for "localhost" port 45592
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/jasmine-2.1.0/lib/jasmine/base.rb:25:in `initialize'
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/jasmine-2.1.0/lib/jasmine/base.rb:25:in `open'
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/jasmine-2.1.0/lib/jasmine/base.rb:25:in `server_is_listening_on'
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/jasmine-2.1.0/lib/jasmine/base.rb:35:in `wait_for_listener'
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/jasmine-2.1.0/lib/jasmine/tasks/jasmine.rake:65:in `block (2 levels) in <top (required)>'
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/airbrake-5.2.0/lib/airbrake/rake/task_ext.rb:19:in `execute'
/home/travis/build/AgileVentures/LocalSupport/vendor/bundle/ruby/2.3.0/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/cli/exec.rb:63:in `load'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/cli/exec.rb:63:in `kernel_load'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/cli/exec.rb:24:in `run'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/cli.rb:304:in `exec'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/cli.rb:11:in `start'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/exe/bundle:27:in `block in <top (required)>'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/lib/bundler/friendly_errors.rb:98:in `with_friendly_errors'
/home/travis/.rvm/gems/ruby-2.3.0@global/gems/bundler-1.12.5/exe/bundle:19:in `<top (required)>'
/home/travis/.rvm/gems/ruby-2.3.0/bin/bundle:23:in `load'
/home/travis/.rvm/gems/ruby-2.3.0/bin/bundle:23:in `<main>'
/home/travis/.rvm/gems/ruby-2.3.0/bin/ruby_executable_hooks:15:in `eval'
/home/travis/.rvm/gems/ruby-2.3.0/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => jasmine:ci
(See full trace by running task with --trace)
[2016-08-10 01:01:42] INFO  going to shutdown ...
[2016-08-10 01:01:42] INFO  WEBrick::HTTPServer#start done.
The command "USE_JASMINE_RAKE=true bundle exec rake jasmine:ci" exited with 1.
```

while it passes locally for me ...

```
[tansaku@Samuels-MacBook-Pro:~/Documents/GitHub/AgileVentures/LocalSupport (119350175_ensure_cuke_failure_leads_to_correct_exit_code)]$ 
→  USE_JASMINE_RAKE=true bundle exec rake jasmine:ci
Waiting for jasmine server on 62907...
Waiting for jasmine server on 62907...
[2016-08-10 17:03:20] INFO  WEBrick 1.3.1
[2016-08-10 17:03:20] INFO  ruby 2.3.0 (2015-12-25) [x86_64-darwin15]
[2016-08-10 17:03:20] INFO  WEBrick::HTTPServer#start: pid=22525 port=62907
Waiting for jasmine server on 62907...
jasmine server started.
..............................

30 specs, 0 failures
[2016-08-10 17:03:23] INFO  going to shutdown ...
[2016-08-10 17:03:23] INFO  WEBrick::HTTPServer#start done.
```

**Comment by tansaku** [comment]
@johnnymo87 sorry, missed your comment before 

> The change to features/support/env.rb is to support a fail-fast option we have configured for cucumber?

the addition of the at_exit to env.rb is the only way I can find to get cucumber to reliably generate a non-zero exit code when single scenarios fail during a batch mode

> The changes to Gemfile.lock and features/step_definitions/volunteer_op_steps.rb are required to make CI pass?

are actually an additional thing here, that perhaps I should remove.  I have been reading of others having this general issue with cucumber, and I thought perhaps an upgrade to the latest version of cucumber-rails might fix things - it didn't, but it did cause an upgrade to a new version of Rails (4.2.7) which then required that one change to volunteer_op_steps.   Having made that upgrade and gotten it working I thought I might as well throw it in here.

Of course now that I'm getting weird jasmine fails I might take them out - it seems like the branch I have is green but the PR itself is red - curiouser and curiouser ...

**Comment by tansaku** [comment]
so the PR is on the merge of my branch into the develop ... hence different commits ... I'm going to try re-ordering when jasmine runs ...

File: features/step_definitions/volunteer_op_steps.rb [hunk 3]
```diff
@@ -25,7 +25,9 @@
 end
 
 Then(/^there should be (\d+) doit volunteer ops stored$/) do |count|
-  expect(VolunteerOp.count(source: 'doit')).to eq count.to_i
+  sql = "SELECT COUNT(*) FROM volunteer_ops WHERE source = 'doit'"
+  doit_ops = VolunteerOp.count_by_sql(sql)
```
Comment (johnnymo87) [inline features/step_definitions/volunteer_op_steps.rb]: Is this the same as
```
VolunteerOp.where(source: 'doit').count
```
?

**Comment by johnnymo87** [comment]
If you are still planning to rebase these changes into the history, it might be worth it to exclude the rails upgrade, because that file is a merge conflict waiting to happen.

File: features/step_definitions/volunteer_op_steps.rb (hunk 3 above)
Comment (tansaku) [inline features/step_definitions/volunteer_op_steps.rb]: yes, probably - it was `VolunteerOp.count(source: 'doit')` and that was throwing an error, where it worked before ... I just ended up trying to work out what the proper use of count with AR was now and that led me to count_by_sql which is damn ugly

**Comment by tansaku** [comment]
@johnnymo87 you might be right about the merge conflict, but it's mainly Gemfile lock, and I don't think any of the few upcoming PRs touch that so I think we're probably safe.  Still, given that I'm trying to rebase into history, the fewer changes the better ...

**Comment by tansaku** [comment]
ultimately I went with a different PR for this - I had thought to fix with a simple upgrade to Ruby 2.3.1 and actually that might have worked, but the associated jasmine fails made me ultimately do the complete upgrade anyway (although perhaps travis has now rolled back what was causing that) - but good to be on more recent libs across the board (perhaps) and anyway, this PR no longer needed ...

**Pull request closed without merging**